### PR TITLE
PMM-12078 HA phase 3 active active PoC

### PR DESCRIPTION
**Pull request by BupycHuk**
PMM-12078

Link to the Feature Build: SUBMODULES-3251

If this PR adds or removes or alters one or more API endpoints, please review and add or update the relevant [API documents](https://github.com/percona/pmm/tree/main/docs/api) as well:

- [ ] API Docs updated

If this PR is related to some other PRs in this or other repositories, please provide links to those PRs:

- Links to related pull requests (optional).

### CODE REVIEW

File: agent/cmd/pmm-agent-entrypoint/main.go [hunk 1]
```diff
@@ -71,15 +71,15 @@ var (
 var pmmAgentProcessID = 0
 
 func runPmmAgent(ctx context.Context, commandLineArgs []string, restartPolicy restartPolicy, l *logrus.Entry, pmmAgentSidecarSleep int) int {
-	pmmAgentFullCommand := "pmm-admin " + strings.Join(commandLineArgs, " ")
```
Comment (JiriCtvrtka) [inline agent/cmd/pmm-agent-entrypoint/main.go:74]: wow

File: agent/cmd/pmm-agent-entrypoint/main.go [hunk 2]
```diff
@@ -71,15 +71,15 @@ var (
...
+	pmmAgentFullCommand := "pmm-agent " + strings.Join(commandLineArgs, " ")
```
Comment (JiriCtvrtka) [inline agent/cmd/pmm-agent-entrypoint/main.go:74]: Is this safe? What if I will do something like:
pmm-agent | mkdir dddd

Comment (BupycHuk) [reply agent/cmd/pmm-agent-entrypoint/main.go:74]: These arguments can be passed only when you run the container, so if you control the container, then you can run all these commands manually.

Comment (JiriCtvrtka) [reply agent/cmd/pmm-agent-entrypoint/main.go:76]: Ok thanks.

**Review by atymchuk — APPROVED** [review]

File: managed/services/supervisord/supervisord.go [hunk 3]
```diff
@@ -519,6 +525,21 @@ func (s *Service) addPostgresParams(templateParams map[string]interface{}) {
...
 }
 
+func (s *Service) addClusterParams(templateParams map[string]interface{}) {
+	templateParams["HAEnabled"] = s.haParams.Enabled
+	if s.haParams.Enabled {
+		templateParams["GrafanaGossipPort"] = s.haParams.GrafanaGossipPort
+		templateParams["HAAdvertiseAddress"] = s.haParams.AdvertiseAddress
+		nodes := make([]string, len(s.haParams.Nodes))
+		for i, node := range s.haParams.Nodes {
+			nodes[i] = fmt.Sprintf("%s:%d", node, s.haParams.GrafanaGossipPort)
```
Comment (artemgavrilov) [inline managed/services/supervisord/supervisord.go:535]: `net.JoinHostPort`

File: managed/services/highavailability/highavailability.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+		s.runLeaderObserver(ctx)
+	}()
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.runRaftNodesSynchronizer(ctx)
+	}()
+
+	<-ctx.Done()
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply managed/services/highavailability/highavailability.go:193]: Yes, services might be empty at this moment of time

File: managed/services/highavailability/services.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+			s.l.Infoln("Stopping ", ls)
+			ls.Stop()
+			s.rw.Lock()
+			defer s.rw.Unlock()
+			delete(s.running, id)
+		}()
+	}
+}
+
+func (s *services) ServiceAdded() chan struct{} {
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply managed/services/highavailability/services.go:101]: Any ideas what can be better?

File: managed/services/supervisord/supervisord.go (hunk 3 above)
Comment (BupycHuk) [reply managed/services/supervisord/supervisord.go:535]: `net.JoinHostPort` requires additional conversion, not sure if it's better

File: managed/services/ha/highavailability.go [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+	return nil
+}
+
+func (s *Service) runRaftNodesSynchronizer(ctx context.Context) {
+	t := time.NewTicker(5 * time.Second)
+
+	for {
+		select {
+		case event := <-s.nodeCh:
```
Comment (JiriCtvrtka) [inline managed/services/ha/highavailability.go:207]: Should we check for nil there?

**Review by JiriCtvrtka — APPROVED** [review]
LGTM, just one small comment.

**Pull request merged**